src/turbine/turbine_tree.zig

```diff
@@ -0,0 +1,910 @@
+            fanout,
+            my_index,
+            shuffled_nodes.items,
+        );
+
+        // Compute the root distance
+        const root_distance: usize = if (my_index == 0)
+            0
+        else if (my_index <= fanout)
+            1
+        else if (my_index <= (fanout +| 1) *| fanout) // Does this make sense?
+            2
+        else
+            3;
+
+        return .{ root_distance, children };
+    }
+
+    /// Create a seeded RNG for the given leader and shred id.
+    /// The resulting RNG must be identical to the agave implementation
+    /// to ensure that the weighted shuffle is deterministic.
+    fn getSeededRng(leader: Pubkey, shred: ShredId) ChaChaRng {
+        const seed = shred.seed(leader);
+        return ChaChaRng.fromSeed(seed);
+    }
+
+    // root     : [0]
+    // 1st layer: [1, 2, ..., fanout]
+    // 2nd layer: [[fanout + 1, ..., fanout * 2],
+    //             [fanout * 2 + 1, ..., fanout * 3],
+    //             ...
+    //             [fanout * fanout + 1, ..., fanout * (fanout + 1)]]
+    // 3rd layer: ...
+    // ...
+    // The leader node broadcasts shreds to the root node.
+    // The root node retransmits the shreds to all nodes in the 1st layer.
+    // Each other node retransmits shreds to fanout many nodes in the next layer.
+    // For example the node k in the 1st layer will retransmit to nodes:
+    // fanout + k, 2*fanout + k, ..., fanout*fanout + k
+    fn computeRetransmitChildren(
+        allocator: std.mem.Allocator,
+        fanout: usize,
+        index: usize,
+        nodes: []const Node,
+    ) !std.ArrayList(Node) {
+        var children = try std.ArrayList(Node).initCapacity(allocator, fanout);
+
+        const offset = (index -| 1) % fanout;
+        const anchor = index - offset;
+        const step = if (index == 0) 1 else fanout;
+        var curr = anchor * fanout + offset + 1;
+        var steps: usize = 0;
+
+        while (curr < nodes.len and steps < fanout) {
+            children.appendAssumeCapacity(nodes[curr]);
+            curr += step;
+            steps += 1;
+        }
+
+        return children;
+    }
+
+    // Returns the parent node in the turbine broadcast tree.
+    // Returns None if the node is the root of the tree.
+    fn computeRetransmitParent(
+        fanout: usize,
+        index_: usize,
+        nodes: []const Node,
+    ) ?Pubkey {
+        var index = index_;
+        const offset = (index -| 1) % fanout;
+        index = if (index == 0) return null else (index - 1) / fanout;
+        index = index - (index -| 1) % fanout;
+        index = if (index == 0) index else index + offset;
+        return nodes[index].pubkey();
+    }
+
+    /// All staked nodes + other known tvu-peers + the node itself;
+    /// sorted by (stake, pubkey) in descending order.
+    fn getNodes(
+        allocator: std.mem.Allocator,
+        my_contact_info: ThreadSafeContactInfo,
+        tvu_peers: []const ThreadSafeContactInfo,
```

Comment:
tvu_peers not really accurate - can just be peers since we filter on shred version -- maybe gossip_peers ?

Reply:
eba48cca

Reply:
8b904c5d